modules/core/arrow-test/src/main/kotlin/arrow/test/laws/SemialignLaws.kt

```diff
@@ -0,0 +1,170 @@
+import arrow.core.extensions.list.functorFilter.flattenOption
+import arrow.core.extensions.list.monadFilter.filterMap
+import arrow.core.toT
+import arrow.typeclasses.Eq
+import arrow.typeclasses.Foldable
+import arrow.typeclasses.Semialign
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+
+object SemialignLaws {
+
+  val iorEq1 = Ior.eq(Int.eq(), Int.eq())
+  val iorEq2 = Ior.eq(Int.eq(), iorEq1)
+
+  fun <F> laws(
+    SA: Semialign<F>,
+    gen: Gen<Kind<F, Int>>,
+    buildEq: (Eq<*>) -> Eq<Kind<F, *>>,
+    FOLD: Foldable<F>
+  ): List<Law> = listOf(
+    Law("Semialign Laws: commutativity") { SA.semialignCommutativity(gen, buildEq(iorEq1)) },
+    Law("Semialign Laws: associativity") { SA.semialignAssociativity(gen, buildEq(iorEq2)) },
+    Law("Semialign Laws: with") { SA.semialignWith(gen, buildEq(String.eq())) },
+    Law("Semialign Laws: functoriality") { SA.semialignFunctoriality(gen, buildEq(Ior.eq(String.eq(), String.eq()))) },
+    Law("Semialign Laws: alignedness") { SA.semialignAlignedness(gen, FOLD) }
+  )
+
+  // Laws ported from https://hackage.haskell.org/package/semialign-1.1/docs/Data-Semialign.html
+
+  fun <F> Semialign<F>.semialignIdempotency() {
+    TODO()
+  }
+
+  fun <F, A> Semialign<F>.semialignCommutativity(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, Ior<A, A>>>) =
+    forAll(G, G) { a: Kind<F, A>, b: Kind<F, A> ->
+      val left: Kind<F, Ior<A, A>> = align(a, b).map { it.swap() }
+      val right: Kind<F, Ior<A, A>> = align(b, a)
+      left.equalUnderTheLaw(right, EQ)
+    }
+
+  fun <F, A> Semialign<F>.semialignAssociativity(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, Ior<A, Ior<A, A>>>>) =
+    forAll(G, G, G) { x: Kind<F, A>, y: Kind<F, A>, z: Kind<F, A> ->
+
+      val left = align(x, align(y, z))
+
+      val right = align(align(x, y), z).map { it.assoc() }
+
+      left.equalUnderTheLaw(right, EQ)
+    }
+
+  fun <F, A> Semialign<F>.semialignWith(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, String>>) =
+    forAll(G, G) { a: Kind<F, A>, b: Kind<F, A> ->
+      val left = alignWith({ "$it" }, a, b)
+      val right = align(a, b).map { "$it" }
+
+      left.equalUnderTheLaw(right, EQ)
+    }
+
+  fun <F, A> Semialign<F>.semialignFunctoriality(G: Gen<Kind<F, A>>, EQ: Eq<Kind<F, Ior<String, String>>>) =
+    forAll(G, G) { a: Kind<F, A>, b: Kind<F, A> ->
+
+      val left = align(a.map { "$it" }, b.map { "$it" })
+      val right = align(a, b).map { ior ->
+        ior.bimap({ "$it" }, { "$it" })
+      }
+
+      left.equalUnderTheLaw(right, EQ)
+    }
+
+  fun <F, A> Semialign<F>.semialignAlignedness(
+    G: Gen<Kind<F, A>>,
+    FOLD: Foldable<F>
+  ) = forAll(G, G) { a: Kind<F, A>, b: Kind<F, A> ->
+
+    fun <E> toList(es: Kind<F, E>): List<E> = FOLD.run {
+      es.foldLeft(emptyList()) { acc, e ->
+        acc + e
+      }
+    }
+
+    val left: List<A> = toList(a)
+
+    // toListOf (folded . here) (align x y)
+    val middle: List<A> = toList(align(a, b).map { it.justLeft() }).flattenOption()
```

Comment:
`flattenOption` is just `filterMap(::identity)` right? So just change `map` to `filterMap` and drop it would work the same I think

Reply:
Oh wait that requires `FunctorFilter`, so this is perfectly fine